lib/cc/cli/test.rb

```diff
@@ -0,0 +1,233 @@
+
+        system([
+          "unset CODE_PATH &&",
+          "unset FILESYSTEM_DIR &&",
+          codeclimate_path,
+          "analyze",
+          "--engine", @engine_name,
+          "-f", "json",
+          relative_path
+        ].join(" "))
+      end
+
+      def prepare_working_dir
+        `git init`
+
+        File.open(".codeclimate.yml", "w") do |config|
+          config.write("engines:\n  #{@engine_name}:\n    enabled: true")
+        end
+      end
+
+      def markers_in(test_file)
+        lines = File.readlines(test_file)
+
+        Array.new.tap do |markers|
+          lines.each_with_index do |line, index|
+            if line =~ /\[issue\].*/
+              markers << build_marker(test_file, index + 1, line)
+            end
+          end
+        end
+      end
+
+      def build_marker(test_file, line_number, text)
+        marker = Marker.new(line_number, text)
+
+        text = text.sub(/^.*\[issue\] ?/, "")
+
+        if text.blank?
+          attrs = {}
+        else
+          matches = text.scan(/([a-z\._-]+)=(?:(")((?:\\.|[^"])*)"|([^\s]*))/).map(&:compact)
+
+          key_values = matches.map do |match|
+            # puts match.inspect
+
+            if match.size == 3 # Quoted
+              key, _, value = match
+              value = '"' + value + '"'
+            else
+              key, value = match
+            end
+
+            [key, munge(value)]
+          end
+
+          attrs = Hash[key_values]
+        end
+
+        issue_line = line_number + 1
+
+        marker.issue = attrs.merge(
+          "engine_name" => @engine_name,
+          "location" => {
+            "path" => test_file,
+            "lines" => {
+              "begin" => issue_line,
+              "end" => issue_line
+            }
+          }
+        )
+
+        if marker.issue["category"]
+          marker.issue["categories"] = Array.wrap(marker.issue["category"])
+          marker.issue.delete("category")
+        end
+
+        marker
+      end
+
+      def munge(value)
+        JSON.load(value)
+      rescue JSON::ParserError
+        value
+      end
+
+      def create_tmpdir
+        tmpdir = File.join("/tmp/cc", SecureRandom.uuid)
+        FileUtils.mkdir_p(tmpdir)
+        tmpdir
+      end
+
+      def unpack(path)
+        system("docker cp #{null_container_id}:#{path} .")
+      end
+
+      def null_container_id
+        # docker cp only works with containers, not images so
+        # hack it by creating a throwaway container
+        @null_container_id = `docker run -d #{engine_image} false`.chomp
+      end
+
+      def remove_null_container
+        `docker rm -f #{null_container_id}` if null_container_id
+      end
+
+      def test_paths
+        Array.wrap(engine_spec["test_paths"])
+      end
+
+      def engine_spec
+        @engine_spec ||= JSON.parse(`docker run --rm #{engine_image} cat /engine.json`)
```

Comment:
I don't see any error handling for if this is missing right now: seems worth adding since not all engines are going to have this file immediately.